caffe2/CMakeLists.txt

```diff
@@ -46,42 +75,93 @@ exclude(Caffe2_DIRS "${Caffe2_DIRS}" ${Caffe2_DIR_EXCLUDES})
+endforeach()
 exclude(Caffe2_GPU_SRCS "${Caffe2_GPU_SRCS}" ${Caffe2_FILE_EXCLUDES})
+exclude(Caffe2_GPU_SRCS "${Caffe2_GPU_SRCS}" ${Caffe2_GPU_TEST_SRCS})
 
 # CPU files.
 foreach(dir ${Caffe2_DIRS})
         file(GLOB tmp ${dir}/*.cc)
         set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS} ${tmp})
 endforeach()
 exclude(Caffe2_CPU_SRCS "${Caffe2_CPU_SRCS}" ${Caffe2_FILE_EXCLUDES})
-exclude(Caffe2_CPU_SRCS "${Caffe2_CPU_SRCS}" ${Caffe2_CPU_FILE_EXCLUDES})
 exclude(Caffe2_CPU_SRCS "${Caffe2_CPU_SRCS}" ${Caffe2_GPU_SRCS})
-
-# CPU test files.
-set(Caffe2_CPU_TEST_SRCS
-        "/test/caffe2_gtest_main.cc"
-)
-prepend(Caffe2_CPU_TEST_SRCS ${CMAKE_CURRENT_SOURCE_DIR} ${Caffe2_CPU_TEST_SRCS})
+exclude(Caffe2_CPU_SRCS "${Caffe2_CPU_SRCS}" ${Caffe2_CPU_TEST_SRCS})
+exclude(Caffe2_CPU_SRCS "${Caffe2_CPU_SRCS}" ${Caffe2_GPU_TEST_SRCS})
 
 # Compile protobufs.
 add_subdirectory(proto)
 include_directories(BEFORE ${CMAKE_BINARY_DIR})
+list(APPEND Caffe2_LINK_LIBS Caffe2_PROTO)
 
 # Compile exposed libraries.
+LIST(APPEND CMAKE_CXX_FLAGS "-std=c++11 -fPIC")
 add_library(Caffe2_CPU ${Caffe2_CPU_SRCS})
-target_link_libraries(Caffe2_CPU Caffe2_PROTO)
+target_link_libraries(Caffe2_CPU ${Caffe2_LINK_LIBS})
+list(APPEND Caffe2_LINK_LIBS Caffe2_CPU)
+
+# CUDA library
+if(HAVE_CUDA)
+  CUDA_ADD_LIBRARY(Caffe2_GPU ${Caffe2_GPU_SRCS})
+  target_link_libraries(Caffe2_GPU ${Caffe2_LINK_LIBS} glog gflags atlas cblas)
+  list(APPEND Caffe2_LINK_LIBS Caffe2_GPU)
+endif()
 
 # Compile test binaries.
-add_executable(Caffe2_CPU_TEST ${Caffe2_CPU_TEST_SRCS})
+foreach(test_src ${Caffe2_CPU_TEST_SRCS})
+  get_filename_component(test_name ${test_src} NAME_WE)
+  add_executable(${test_name} "${test_src}")
+  target_link_libraries(${test_name} ${Caffe2_LINK_LIBS} glog atlas cblas gtest gtest_main ${Caffe2_LINKER_LIBS})
+  target_compile_features(${test_name} PRIVATE cxx_range_for)
+endforeach()
+
+# ---[ Python
+add_subdirectory(python)
+
+# ---[ All binaries
+file(GLOB Caffe2_BINARY_SRCS ${CMAKE_CURRENT_SOURCE_DIR}/binaries/*.cc)
+
+# ZMQ src
+set(Caffe2_ZMQ_BINARY_SRC
+    "zmq_feeder.cc"
+)
+# Prune out ZMQ if no support
+if(NOT USE_ZMQ)
+  message(STATUS "Excluding ZMQ binary")
+  prepend(tmp "${CMAKE_CURRENT_SOURCE_DIR}/binaries/" "${Caffe2_ZMQ_BINARY_SRC}")
+  exclude(Caffe2_BINARY_SRCS "${Caffe2_BINARY_SRCS}" "${tmp}")
+endif()
+
+# MPI-based binaries
+set(Caffe2_MPI_BINARY_SRC
+    "fb_run_plan_mpi.cc"
+    "run_plan_mpi.cc"
+)
+if(NOT USE_MPI)
+  message(STATUS "Excluding MPI binaries")
+  prepend(tmp "${CMAKE_CURRENT_SOURCE_DIR}/binaries/" "${Caffe2_MPI_BINARY_SRC}")
+  exclude(Caffe2_BINARY_SRCS "${Caffe2_BINARY_SRCS}" "${tmp}")
+endif()
+
+foreach(binary_src ${Caffe2_BINARY_SRCS})
+  get_filename_component(bin_name ${binary_src} NAME_WE)
+  add_executable(${bin_name} ${binary_src})
+  target_link_libraries(${bin_name} -Wl,--whole-archive ${Caffe2_LINK_LIBS} glog atlas cblas cnmem ${Caffe2_LINKER_LIBS} -Wl,--no-whole-archive)
```

Comment:
`atlas` and `cnmem` are optional, so they may not even be in this compilation path